Added benchmark of anelastic Vs to temperature conversion

freddrichards opened the pull request:
I've added a new benchmark for the anelastic Vs to temperature conversion and updated the manual accordingly.

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/yamauchi_takei_2016_anelasticity.prm:
```diff
@@ -0,0 +1,90 @@
+  set Use Yamauchi and Takei parameterization = true
+  set Use original density and frequency model of Yamauchi and Takei = true
+  subsection Ascii data model
+  	set Data directory       = /Users/fredrichards/aspect_master/data/initial-temperature/ascii-data/test/
```
Can you use ``$ASPECT_SOURCE_DIR/data/`` here?

freddrichards replied:
Yes, have changed

tjhei commented on doc/manual/manual.tex:
```diff
@@ -10432,6 +10432,89 @@ \subsubsection{Donea \& Huerta 2D box geometry benchmark}
+
+\textit{This section was contributed by Fred Richards.}
+
+This benchmark tests the implementation of the anelastic shear wave velocity-to-temperature conversion derived by Yamauchi \& Takei~\cite{YT16} based on forced-oscillation experiments conducted at seismic frequencies on polycrystalline borneol. This anelasticity parameterization has been calibrated against a range of observational constraints on upper mantle temperature, attenuation and viscosity ...
```
Can you add a comment where the user can find the benchmark (benchmarks/..../)?

tjhei commented on doc/manual/manual.tex:
```diff
@@ -10432,6 +10432,89 @@ \subsubsection{Donea \& Huerta 2D box geometry benchmark}
+\begin{figure}
+\begin{center}
+  \centering
+  \includegraphics[width=\textwidth]{cookbooks/benchmarks/anelasticity/YT16_benchmark.jpg}
```
did you put the .jpg files in the correct place? (doc/manual/cookbooks/benchmarks/...)

freddrichards replied:
these are in the correct place as far as I can tell (compiled correctly on my machine), however,  I have changed the file to a .png to make the file size a bit smaller.

tjhei commented on doc/manual/manual.tex:
```diff
@@ -10432,6 +10432,89 @@ \subsubsection{Donea \& Huerta 2D box geometry benchmark}
+7& \text{for } T^{\prime} \geq 1 \\
+\end{cases}
+\end{equation}
+where $T^{\prime}$ is homologous temperature ($\frac{T}{T_{s}}$) with $T$ the temperature and $T_s$ the solidus temperature, both in Kelvin. $\phi_m$ is the melt fraction and $\beta(\phi_m)$ describes the direct poroelastic effect of melt (assumed to be negligible under upper mantle conditions). For this case, $J_U$ is the inverse of the unrelaxed shear modulus, $\mu_{U}(P,T)$, such that
```
Can you check the grammar of this sentence? "is the"?

tjhei commented on doc/manual/manual.tex:
```diff
@@ -10432,6 +10432,89 @@ \subsubsection{Donea \& Huerta 2D box geometry benchmark}
+\gamma^{-1} \text{exp}(\lambda\phi) & \text{for }T^{\prime} \geq 1\\
+\end{cases}
+\end{equation}
+where $T^{\prime}_\eta$ is the homologous temperature above which activation energy becomes $E_a + \Delta E_a$ and $\gamma = 5$ is the factor of additional reduction. $\lambda\phi$ describes the direct effect of melt on viscosity, assumed to be negligible here. The solidus temperature, $T_s$, is fixed to a value of 1599 K at 50~km equivalent to a dry peridotite solidus \cite{Hirsch2000} and linea ...
```
, which ?

freddrichards replied:
"above which" is grammatically correct but I have put a comma before "and $\gamma..." so it's clear that the statements about activation energy and $\gamma$ are separate.

tjhei commented on doc/manual/manual.tex:
```diff
@@ -10432,6 +10432,89 @@ \subsubsection{Donea \& Huerta 2D box geometry benchmark}
+
+In this benchmark, a 2D input ASCII file containing $V_S$ specified at 50 and 75 km depth is read in and converted to temperature using an initial temperature model that implements the $V_{S}(P,T)$ formulation detailed above. The parameters governing the relationship between $V_S$ and temperature are set to the values calibrated by Yamauchi \& Takei~\cite{YT16}, where $\mu_U^0 = 72.45$~GPa, $\fra ...
+
+The parameter file and initial temperature model for this benchmark can be found at \url{benchmarks/yamauchi_takei_2016_anelasticity/benchmarks/yamauchi_takei_2016_anelasticity.prm} and \url{benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc}. Code to recreate  Fig.~\ref{fig:anelasticity}  is provided in \url{benchmarks/yamauchi_takei_2016_anelasticity/benchmarks/plot_output} ...
```
The path of the prm seems wrong

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+    AnelasticVs2T<dim>::
+    fVs (double x,
+	     const double      depth,
+		 const double      absolute_Vs,
```
indentation looks wrong. Did you run ./contrib/utilities/indent ?

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+	  // initialize depth
+      double depth;
+      // if working in a box geometry, take y coordinate
+      if (this->get_geometry_model().has_curved_elements() == 0){depth = position(1);}
```
use false instead of 0, but braces on extra line.

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+      double depth;
+      // if working in a box geometry, take y coordinate
+      if (this->get_geometry_model().has_curved_elements() == 0){depth = position(1);}
+      // if working in a spherical geometry, take find depth from geometry model call
```
why does this not work for the box?

freddrichards replied:
in the box example I was using the y coordinate as depth, however, the get_geometry_model().depth(position) call seems to expect the y coordinate to be radius meaning that depth gets read in upside down if this option is used. I feel that having the y coordinate equal to depth in the input file is more intuitive but if that's not consistent with other models then I can change this.

bangerth replied:
But the way you have it, the `depth` you compute here *increases* from bottom to top! That can't be what you wanted, right?

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+      // if working in a box geometry, take y coordinate
+      if (this->get_geometry_model().has_curved_elements() == 0){depth = position(1);}
+      // if working in a spherical geometry, take find depth from geometry model call
+      else{depth = this->get_geometry_model().depth(position);}
```
put braces on extra lines never on the same line with the code.

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+	  const double a=1;
+	  const double b=3;
+	  const double tol=1.48e-8;
+	  int const max_iter=500;
```
const int
```suggestion
	  const int max_iter=500;
```

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+		  *(1/temperature-1/reference_temperature))*exp((activation_volume/gas_constant)*(pressure/temperature-reference_pressure/
+			  reference_temperature))*viscosity_reduction_factor;
+	  unrelaxed_compliance=1/(1e9*(mu0+(dmudP*pressure*1e-9)+(dmudT*(temperature-273))));
+	  if ((temperature-273<0) || (viscosity > 1e40))
```
temperature < 273.0

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+	  unrelaxed_compliance=1/(1e9*(mu0+(dmudP*pressure*1e-9)+(dmudT*(temperature-273))));
+	  if ((temperature-273<0) || (viscosity > 1e40))
+	    {
+		  temperature=273;
```
you want to reset both even if one is invalid?

freddrichards replied:
I have changed this now, I only really want to change these things if temperature is less than 273

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+		  // calculate shear wave period incorporating depth dependence of Forsyth 1992
+	      period=(3*depth)/4200;
+	    }
+	  else
```
flip if and else here

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+	    double fs = 0;      // initialize
+	    if (!(fa * fb < 0))
+	    {
+	        std::cout << "Signs of f(lower_bound) and f(upper_bound) must be opposites" << std::endl; // throws exception if root isn't bracketed
```
this comment is incorrect. Do you want to throw here?

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+	            -------------------------------------------------------
+	            (condition 1) s is not between  (3a+b)/4  and b or
+	            (condition 2) (mflag is true and |s−b| ≥ |b−c|/2) or
+	            (condition 3) (mflag is false and |s−b| ≥ |c−d|/2) or
```
is this code taken from somewhere?

freddrichards replied:
I've now replaced this Brent function with the boost version so this issue should now be fixed

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,469 @@
+        
+		Utilities::AsciiDataBase<dim>::declare_parameters(prm,
+                         "$ASPECT_SOURCE_DIR/data/initial-temperature/ascii-data/test/",
+                         "box_2d.txt");
```
is this a sensible default?

freddrichards replied:
I have changed this to box_2d_Vs_YT16.txt which is the 2D example Vs input used in the benchmark.

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.h:
```diff
@@ -0,0 +1,136 @@
+       * Brent minimization function to find temperature from Vs and isothermal compression from pressure
+       */
+       static
+	   double
```
this function has little to do with your class here. Can you move it into the .cc file as a static function?

tjhei commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.h:
```diff
@@ -0,0 +1,136 @@
+	   * Function that assesses difference between input and calculated Vs for Brent minimization
+	   */
+	   double
+	   fVs(double x, double depth, double absolute_Vs, double mu0, double dmudT,
```
you can probably make these functions private.

gassmoeller commented:
/rebuild

gassmoeller commented:
You will need to update the indentation, if you have never done that talk to one of us.

freddrichards commented:
I fixed the code so it calls the brent function from boost/math/tools/minima.hpp and addressed other comments. Responses to things requiring more discussion are included above.

tjhei commented:
Due to the change of the location of the indent script in the last few days, you will need to rebase to the current master. Sorry about that.

tjhei commented:
Please remove the .DS_Store file from this PR.

tjhei commented:
Can you please rerun ``./contrib/utitlies/indent``? It seems that you have tabs in your .prm.

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,408 @@
+         const double      solidus_gradient,
+         const bool      use_original_model) const
+    {
+      return abs(yamauchi_takei_Vs(x,depth,mu0,dmudT,dmudP,viscosity_prefactor,activation_energy,activation_volume
```
You'll have to write this as `std::abs` for complicated and historical reasons :-( Maybe the same in other places below -- please do a search-replace.

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,408 @@
+    AnelasticVs2T<dim>::
+    fdV (double x, const double bulk_modulus, const double bulk_modulus_pressure_derivative, const double pressure ) const
+    {
+      return abs((bulk_modulus*(3./2.)*(pow(x,7./3.)-pow(x,5./3.))*(1+(((3./4.)*
```
like here

While you're there, you might also want to write `std::pow`, even though the existing form is not technically wrong (as it is for `abs`).

freddrichards replied:
Done

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,408 @@
+      // initialize depth
+      double depth;
+      // if working in a box geometry, take y coordinate
+      if (this->get_geometry_model().has_curved_elements() == false)
```
I'm not sure this is a reliable way to do this -- just always use `geometry_model.depth(...)` as you do in the `else` branch.

freddrichards replied:
I've edited the relevant files to make the 2D box work with "geometry_model.depth(...)"

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,408 @@
+        }
+
+      // initialize temperature
+      double temperature;
```
actually, you specifically *don't* initialize the variable here :-) You just declare it.

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,408 @@
+          attenuation=1e-9;
+
+        }
+      // evaluate Maxwell normalised shear wave period
```
no need for the empty line above the `}`. but you're starting something new after the brace, so maybe have an empty line before the comment?

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc:
```diff
@@ -0,0 +1,408 @@
+      // determine J1 term (real part of complex compliance)
+      storage_compliance=unrelaxed_compliance*(1+((background_amplitude*pow(normalised_period,background_slope))
+                                                  /background_slope)+((sqrt(2*M_PI)/2)*peak_amplitude*peak_width*(1-
+                                                                      erf((log(peak_period/normalised_period))/(sqrt(2)*peak_width)))));
```
`erf`, `pow`, `sqrt` and `exp` should all be prefixed by `std::`

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.h:
```diff
@@ -0,0 +1,128 @@
+
+    /**
+     * A class that implements a prescribed temperature field determined from
+     * a AsciiDataProfile input file.
```
I think you got this comment by copying from somewhere. Can you update it?

freddrichards replied:
Done

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.h:
```diff
@@ -0,0 +1,128 @@
+      public:
+        /**
+        * Constructor. Initialize variables.
+        */
```
there's something funny going on with the alignment of the asterisks

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.h:
```diff
@@ -0,0 +1,128 @@
+        double
+        yamauchi_takei_Vs(double temperature, double depth, double mu0, double dmudT,
+                          double dmudP, double viscosity_prefactor, double activation_energy,
+                          double activation_volume, double solidus_gradient, bool density_model_flag) const;
```
you can probably put this function into the `private` section of the class.

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/plot_output/gmt.history:
```diff
@@ -0,0 +1,8 @@
+# GMT 5 Session common arguments shelf
+BEGIN GMT 5.4.4
+B	px100+lTemperature (@~\260@~C)py0.1f0.01+lV@-S@- (km s@+-1@+)WSne
```
That's bizarre :-)

bangerth commented on benchmarks/yamauchi_takei_2016_anelasticity/plot_output/gmt.conf:
```diff
@@ -0,0 +1,11 @@
+#
+# GMT 5.4.4 Defaults file
```
Will those who run your benchmark need to use these GMT files?

freddrichards replied:
Nope, have removed

bangerth commented on doc/manual/manual.tex:
```diff
@@ -10501,6 +10501,91 @@ \subsubsection{Advection Stabilization Benchmarks}
+
+\textit{This section was contributed by Fred Richards.}
+
+This benchmark tests the implementation of the anelastic shear wave velocity-to-temperature conversion derived by Yamauchi \& Takei~\cite{YT16} based on forced-oscillation experiments conducted at seismic frequencies on polycrystalline borneol. This anelasticity parameterization has been calibrated against a range of observational constraints on upper mantle temperature, attenuation and viscosity ...
```
```suggestion
This benchmark tests the implementation of the anelastic shear wave velocity-to-temperature conversion derived by Yamauchi \& Takei~\cite{YT16} based on forced-oscillation experiments conducted at seismic frequencies on polycrystalline borneol. This anelasticity parameterization has been calibrated against a range of observational constraints on upper mantle temperature, attenuation and viscosity structure, using the surface wave tomography model of Priestley et al.~\cite{P12} to constrain shear wave velocity ($V_S$) and the plate model of \cite{McK05} to estimate lithospheric temperature structure. The resulting $V_S$-to-temperature conversion accurately accounts for the strongly non-linear temperature dependence of $V_S$ at near-solidus conditions and is therefore especially useful for initializing models with accurate temperature structure in the upper $\sim$~400 km of the mantle. This benchmark is located in the folder \url{benchmarks/yamauchi_takei_2016_anelasticity}.
```

bangerth commented on doc/manual/manual.tex:
```diff
@@ -10501,6 +10501,91 @@ \subsubsection{Advection Stabilization Benchmarks}
+\begin{center}
+  \centering
+  \includegraphics[width=\textwidth]{cookbooks/benchmarks/anelasticity/YT16_benchmark.png}
+  \caption{\it $V_S$ as a function of temperature in the oceanic lithosphere. Dotted lines = digitized results from Fig. 20 of Yamauchi \& Takei~\cite{YT16}; solid lines = \aspect{} results; red = 50 km; blue = 75 km. Temperatures are taken from the plate model of McKenzie et al.~\cite{McK05} and $V_{S}$ from the surface wave tomography model of Priestley et al.\cite{P12}.}  
```
```suggestion
  \caption{\it $V_S$ as a function of temperature in the oceanic lithosphere. Dotted lines: digitized results from Fig. 20 of Yamauchi \& Takei~\cite{YT16}; solid lines: \aspect{} results; red = 50 km; blue = 75 km. Temperatures are taken from the plate model of McKenzie et al.~\cite{McK05} and $V_{S}$ from the surface wave tomography model of Priestley et al.~\cite{P12}.}  
```

bangerth commented on doc/manual/manual.tex:
```diff
@@ -10501,6 +10501,91 @@ \subsubsection{Advection Stabilization Benchmarks}
+\rho (P,T) = \rho_{0}  \left\{ 1 - \left[\alpha(T - T_0)\right] + \frac{P}{K} \right\}
+\label{eq:density_P}
+\end{equation}
+where $rho_{0} = 3291$~kg~m$^{-3}$ and $\alpha = 3.59 \times 10^{-5}$~K$^{-1}$ are the density and thermal expansivity corresponding to $T_{0} = 873$~K, $P$ = pressure and $K = 115.2$~GPa is the bulk modulus. $J_1$ and $J_2$ are expressed as
```
```suggestion
where $rho_{0} = 3291\, \text{kg}\ \text{m}$^{-3}$ and $\alpha = 3.59 \times 10^{-5}\text{K}^{-1}$ are the density and thermal expansivity corresponding to $T_{0} = 873\text{K}$, $P$ is the pressure and $K = 115.2\text{GPa}$ is the bulk modulus. $J_1$ and $J_2$ are expressed as
```

bangerth commented on doc/manual/manual.tex:
```diff
@@ -10501,6 +10501,91 @@ \subsubsection{Advection Stabilization Benchmarks}
+J_1(\tau_S^{\prime})= J_U \left[ 1 + \frac{A_B[\tau_S^{\prime}]^{\alpha_B}}{\alpha_B} + \frac{\sqrt{2\pi}}{2} A_P~\sigma_P \left\{ 1-\text{erf}\left(\frac{\text{ln}[\tau_P^{\prime}/\tau_S^{\prime}]}{\sqrt{2}\sigma_P}\right)\right\}\right]
+\label{eq:real_compliance}
+\end{equation}
+\begin{equation}
```
I think it might be nicer to use the `{align}` environment and align the two equations for J1 and J2.

freddrichards replied:
Done

bangerth commented on doc/manual/manual.tex:
```diff
@@ -10501,6 +10501,91 @@ \subsubsection{Advection Stabilization Benchmarks}
+J_2(\tau_S^{\prime}) = J_U \frac{\pi}{2} \left[ A_B[\tau_S^{\prime}]^{\alpha_B} + A_P~\text{exp} \left(-\frac{\text{ln}^2[\tau_P^{\prime}/\tau_S^{\prime}]}{2\sigma_P^2}\right)\right] + J_U \tau_S^{\prime}
+\label{eq:imaginary_compliance}
+\end{equation}
+where $A_B$ = 0.664 and $\alpha_B$ = 0.38 represent the amplitude and slope of background stress relaxation and $J_U$ is the unrelaxed compliance. Parameters $A_P$ and $\sigma_P$ represent the amplitude and width of a high frequency relaxation peak superimposed on this background trend such that
```
```suggestion
where $A_B = 0.664$ and $\alpha_B = 0.38$ represent the amplitude and slope of background stress relaxation and $J_U$ is the unrelaxed compliance. Parameters $A_P$ and $\sigma_P$ represent the amplitude and width of a high frequency relaxation peak superimposed on this background trend such that
```

bangerth commented on doc/manual/manual.tex:
```diff
@@ -10501,6 +10501,91 @@ \subsubsection{Advection Stabilization Benchmarks}
+\end{equation}
+where $\frac {\partial T_s}{\partial z}$ is the solidus gradient. 
+
+In this benchmark, a 2D input ASCII file containing $V_S$ specified at 50 and 75 km depth is read in and converted to temperature using an initial temperature model that implements the $V_{S}(P,T)$ formulation detailed above. The parameters governing the relationship between $V_S$ and temperature are set to the values calibrated by Yamauchi \& Takei~\cite{YT16}, where $\mu_U^0 = 72.45$~GPa, $\fra ...
```
similar issues here with expressing units as mentioned above. (See also #3044 if you want to do it "right".)

freddrichards replied:
This should now be sorted.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -10501,6 +10501,88 @@ \subsubsection{Advection Stabilization Benchmarks}
+
+In this benchmark, a 2D input ASCII file containing $V_S$ specified at 50 and 75 km depth is read in and converted to temperature using an initial temperature model that implements the $V_{S}(P,T)$ formulation detailed above. The default parameters governing the relationship between $V_S$ and temperature are set to the values calibrated by Yamauchi \& Takei~\cite{YT16}, where $\mu_U^0 = 72.45~\te ...
+
+The parameter file and initial temperature model for this benchmark can be found at \url{benchmarks/yamauchi_takei_2016_anelasticity/yamauchi_takei_2016_anelasticity.prm} and \url{benchmarks/yamauchi_takei_2016_anelasticity/anelasticity_temperature.cc}. Code to recreate  Fig.~\ref{fig:anelasticity}  is provided in \url{benchmarks/yamauchi_takei_2016_anelasticity/plot_output}.
```
@cedrict : FYI -- this is how one can reference individual files.

bangerth commented:
I merged too early: Please also add another pull request that adds an entry to `doc/modules/changes` that lists the new benchmark.
